File: tests/engine/request-support.ts
/**
 * The chart every test of a read runs on, and the two helpers that drive it.
 *
 * Shared by `requests.test.ts` and `request-settings.test.ts` rather than
 * written twice, because the numbers each of them asserts are numbers about
 * these bars: five minute bars priced by index, two whole hourly buckets, dated
 * at instants a reader can check against a clock. A second copy of the fixture
 * would be a second set of boundaries to keep in step, and an off-by-one in one
 * of them would look like a disagreement between the two files.
 */
import assert from 'node:assert/strict';

import { load } from '../../src/core/engine/index.js';
import type { EngineHost, HostBar, Value } from '../../src/core/engine/index.js';
import { asWire, compile } from './support.js';

/** A chart of five minute bars, priced by index so a value names its bar. */
const OPEN = Date.UTC(2025, 0, 6, 10, 0, 0);

export function fiveMinutes(count: number): readonly HostBar[] {
  const out: HostBar[] = [];
  for (let i = 0; i < count; i += 1) {
    out.push({
      open: 100 + i,
      high: 100 + i + 0.5,
      low: 100 + i - 0.5,
      close: 100 + i,
      volume: 10,
      time: OPEN + i * 300_000,
    });
  }
  return out;
}

export const CHART: EngineHost = {
  instrument: { symbol: 'AAA', exchange: 'XX', interval: '5', timezone: 'UTC' },
};

/**
 * Compiles, and refuses to go on if the script itself was wrong.
 *
 * A test script with a name that does not exist still produces a program, and
 * the plot it was supposed to drive is then absent for a reason that has nothing
 * to do with what is being tested. Asserting here is what turns that into a
 * failure with the code on it rather than a passing test of nothing.
 */
export function clean(source: string): ReturnType<typeof compile> {
  const compiled = compile('request.oscript', source);
  assert.deepEqual(
    compiled.diagnostics.filter((one) => one.severity === 'error').map((one) => one.code),
    [],
    'the test script itself did not compile',
  );
  return compiled;
}

export function ran(
  source: string,
  count = 24,
  host: EngineHost = CHART,
  settings: Readonly<Record<string, unknown>> = {},
): readonly Value[][] {
  const compiled = clean(source);
  const loaded = load(asWire(compiled.program), { source: compiled.file, host, settings });
  if (!loaded.ok) throw new Error(`${loaded.diagnostic.code}: ${loaded.diagnostic.message}`);
  const bars = fiveMinutes(count);
  const result = loaded.engine.run(bars, bars.map(() => ({ isConfirmed: true })));
  assert.equal(result.diagnostic, undefined, result.diagnostic?.message);
  return loaded.engine.program.channels.map((_, channel) => [...loaded.engine.column(channel)]);
}

File: engine/tests/request_support.py
"""The chart every test of a read runs on, and a builder for programs that read.

Shared by ``test_requests.py`` and ``test_request_host.py`` rather than written
twice, because the numbers each asserts are numbers about these bars: five
minute bars from 10:00 on 6 January 2025, priced by their index so a value in a
column says which bar produced it, twelve to the hour. That is the first
engine's own fixture for the same questions, so the two engines' tests of the
fold read the same boundaries.

The programs are built by hand, with the ``requests`` table section 2.16 prints,
because this package has no compiler and a test that needed one would be a test
of something else. Each read's value lands in a ``"request"`` register and each
register is emitted to a channel of its own, so a column is one read.
"""

from typing import Any, Dict, List, Optional, Sequence

from tests.support import RESERVED, channel, program

from openscript.contracts import Bar, BarState
from openscript.run import load
from openscript.verify import capabilities

#: 10:00 UTC on Monday 6 January 2025.
OPEN = 1736157600000
FIVE_MINUTES = 300_000

#: The chart's own record: five minute bars, dated in UTC.
CHART: Dict[str, Any] = {"symbol": "AAA", "exchange": "XX", "interval": "5", "timezone": "UTC"}


def five_minutes(count: int, start: int = 0) -> List[Bar]:
    """Bars priced by index: open and close ``100 + i``, a half either side, volume 10."""
    return [
        Bar(
            time=float(OPEN + i * FIVE_MINUTES),
            open=100.0 + i,
            high=100.0 + i + 0.5,
            low=100.0 + i - 0.5,
            close=100.0 + i,
            volume=10.0,
        )
        for i in range(start, start + count)
    ]


def body(field: str, code: Optional[Sequence[Any]] = None, requests: Sequence[Any] = (), extra: int = 0) -> Dict[str, Any]:
    """A read's body over one bar field, register 0, answering it unless told otherwise."""
    series = [{"id": 0, "kind": "bar", "field": field, "name": field}]
    series += [{"id": 1 + at, "kind": "request", "field": None, "name": f"r{at}"} for at in range(extra)]
    return {
        "inputs": [],
        "series": series,
        "frame": {"slots": 0},
        "cells": [],
        "states": [],
        "functions": [],
        "callSites": [],
        "loops": [],
        "requests": list(requests),
        "code": [list(one) for one in (code or [["SLOAD", 0], ["RET"]])],
        "pos": [[0, 3, 5]],
        "fnPos": [],
    }


def read(ident: int, timeframe: str, mode: str, series: int, inner: Dict[str, Any], kind: str = "timeframe", symbol: Any = None) -> Dict[str, Any]:
    """One entry of ``requests``, section 2.16's shape."""
    return {
        "id": ident,
        "read": kind,
        "symbol": symbol,
        "exchange": None,
        "timeframe": timeframe,
        "mode": mode,
        "series": series,
        "warmup": None,
        "body": inner,
    }


def reading(requests: Sequence[Dict[str, Any]], lib: Optional[List[Any]] = None, tail: Sequence[Any] = (), consts: Optional[List[Any]] = None) -> Dict[str, Any]:
    """A program that emits each read's register to a channel of its own, then ``tail``."""
    count = len(requests)
    code: List[Any] = []
    for at in range(count):
        code += [["SLOAD", at], ["EMIT", at]]
    code += [list(one) for one in tail]
    extra = sum(1 for one in tail if one[0] == "EMIT")
    tags = {"core.1"} | {"req." + one["read"] for one in _all(requests)}
    return program(
        code + [["HALT"]],
        consts=consts if consts is not None else list(RESERVED),
        channels=[channel(at) for at in range(count + extra)],
        series=[{"id": at, "kind": "request", "field": None, "name": f"r{at}"} for at in range(count)],
        requests=list(requests),
        requires=sorted(tags),
        lib={"manifest": 1, "functions": list(lib or [])},
    )


def _all(requests: Sequence[Dict[str, Any]]) -> List[Dict[str, Any]]:
    found = []
    for one in requests:
        found.append(one)
        found.extend(_all(one["body"]["requests"]))
    return found


def loaded(built: Dict[str, Any], record: Optional[Dict[str, Any]] = None, provider: Any = None, library: Any = None):
    """The program loaded on the chart above, or the load's refusal."""
    return load(
        built,
        {},
        library,
        capabilities=capabilities(),
        instrument=CHART if record is None else record,
        provider=provider,
    )


def columns(built: Dict[str, Any], bars: Sequence[Bar], again: int = 0, history: bool = True, **given: Any) -> List[List[Any]]:
    """Every channel over every bar, each bar executed ``again`` more times on top.

    ``history`` hands the whole dataset to the fold before bar 0, as a run over a
    history does; without it the bars arrive one at a time, as a live feed does.
    """
    result = loaded(built, **given)
    if not result.ok:
        raise AssertionError(f"the program was refused: {result.diagnostic}")
    run = result.run
    if history:
        run.history(bars)
    record = given.get("record") or CHART
    rows = []
    for index, bar in enumerate(bars):
        for _ in range(again + 1):
            out = run.execute_bar(index, bar, BarState(), supplied=len(bars), instrument=record)
            if not out.ok:
                raise AssertionError(f"bar {index} failed: {out.diagnostic}")
        rows.append(out.columns)
    return [list(column) for column in zip(*rows)]
